light-client/src/peer_list.rs

```diff
@@ -5,111 +5,169 @@ use crate::{
     types::PeerId,
 };
 
-use contracts::pre;
-use std::collections::HashMap;
+use contracts::{post, pre};
+use std::collections::{HashMap, HashSet};
 
 /// A mapping from PeerIds to Light Client instances.
 /// Keeps track of which peer is deemed the primary peer.
 #[derive(Debug)]
 pub struct PeerList {
-    peers: HashMap<PeerId, Instance>,
+    instances: HashMap<PeerId, Instance>,
     primary: PeerId,
+    witnesses: HashSet<PeerId>,
+    full_nodes: HashSet<PeerId>,
+    faulty_nodes: HashSet<PeerId>,
 }
 
 impl PeerList {
+    /// Invariant maintained by a `PeerList`
+    ///
+    /// ## Implements
+    /// - [LCD-INV-NODES]
+    pub fn invariant(peer_list: &PeerList) -> bool {
+        peer_list.full_nodes.is_disjoint(&peer_list.witnesses)
+            && peer_list.full_nodes.is_disjoint(&peer_list.faulty_nodes)
+            && peer_list.witnesses.is_disjoint(&peer_list.faulty_nodes)
+    }
+
+    /// Transition invariant maintained by a `PeerList`
+    ///
+    /// ## Implements
+    /// - [LCD-INV-NODES]
+    pub fn transition_invariant(_prev: &PeerList, _next: &PeerList) -> bool {
+        true
+        // TODO
+        // &next.full_nodes | &next.witnesses | &next.faulty_nodes
+        //     == &prev.full_nodes | &prev.witnesses | &prev.faulty_nodes
+    }
+
     /// Returns a builder of `PeerList`
     pub fn builder() -> PeerListBuilder {
         PeerListBuilder::default()
     }
 
     /// Get a reference to the light client instance for the given peer id.
     pub fn get(&self, peer_id: &PeerId) -> Option<&Instance> {
-        self.peers.get(peer_id)
+        self.instances.get(peer_id)
     }
 
     /// Get a mutable reference to the light client instance for the given peer id.
     pub fn get_mut(&mut self, peer_id: &PeerId) -> Option<&mut Instance> {
-        self.peers.get_mut(peer_id)
+        self.instances.get_mut(peer_id)
     }
 
     /// Get a reference to the current primary instance.
     pub fn primary(&self) -> Option<&Instance> {
-        self.peers.get(&self.primary)
+        self.instances.get(&self.primary)
     }
 
     /// Get a mutable reference to the current primary instance.
     pub fn primary_mut(&mut self) -> Option<&mut Instance> {
-        self.peers.get_mut(&self.primary)
+        self.instances.get_mut(&self.primary)
     }
 
     /// Get a list of references to all the witnesses,
     /// (ie. all peers which are not the primary).
     pub fn witnesses(&self) -> Vec<&Instance> {
-        self.peers
+        self.instances
```

Comment:
This method now seems to be wrong as it will return all `Instance` which are not `primary` (as per the docs above but not what we want when doing fork detection), therefore we might be returning faulty nodes, and we are going to contact all instances in this peer list. Effectively communicating with all full nodes instead of a subset defined in `witnesses`. This could probably be fixed in #397 @romac

Reply:
Fixed in https://github.com/informalsystems/tendermint-rs/pull/397/commits/e371271820e11568244b13ada2616e90b63133a9.